File: hackerearth/practice/data-structures/arrays/1-d/distinct-count-2/solution.cpp
// https://www.hackerearth.com/practice/data-structures/arrays/1-d/practice-problems/golf/distinct-count-2/

#include<bits/stdc++.h>
using namespace std;

int main() {
	map<int, int> freq;
	int N;
	cin >> N;

	int maxN = 0, maxFreq = 0;
	while (N--) {
		int x;
		cin >> x;
		
		if (freq.find(x) == freq.end())
			freq[x] = 1;
		else
			freq[x]++;

		if (maxFreq < freq[x]) {
			maxFreq = freq[x];
			maxN = x;
		} else if (maxFreq == freq[x] && x < maxN) {
			maxN = x;
		}
	}

	cout << maxN;

	return 0;
}

File: hackerearth/practice/data-structures/arrays/1-d/distinct-count-2/Solution.java
// https://www.hackerearth.com/practice/data-structures/arrays/1-d/practice-problems/golf/distinct-count-2/

import java.util.*;

class Solution {
    public static void main(String args[] ) throws Exception {
        Scanner input = new Scanner(System.in);
        int N = input.nextInt();
        HashMap<Integer, Integer> freq = new HashMap<>();

        int ans = 0, ansFreq = 0;
        for (int i = 0; i < N; i++) {
            int x = input.nextInt();

            int count = freq.getOrDefault(x, 0) + 1;
            freq.put(x, count);
            if (ansFreq < count) {
                ans = x;
                ansFreq = count;
            } else if (ansFreq == count && x < ans) {
                ans = x;
            }
        }

        input.close();

        System.out.println(ans);
    }
}
